src/Microsoft.Identity.Web.TokenAcquisition/TokenAcquisition.cs

```diff
@@ -585,10 +608,75 @@ private bool IsInvalidClientCertificateOrSignedAssertionError(MsalServiceExcepti
                     _applicationsByAuthorityClientId.TryAdd(GetApplicationKey(mergedOptions), application);
                 }
             }
+            return application;
+        }
 
+        /// <summary>
+        /// Gets a cached ManagedIdentityApplication object or builds a new one if not found.
+        /// </summary>
+        /// <param name="mergedOptions"></param>
+        /// <param name="managedIdentityOptions"></param>
+        /// <returns></returns>
+        internal async Task<IManagedIdentityApplication> GetOrBuildManagedIdentityApplication(
+            MergedOptions mergedOptions, ManagedIdentityOptions managedIdentityOptions)
+        {
+            IManagedIdentityApplication? application;
+            string key = GetCacheKeyForManagedId(managedIdentityOptions);
+
+            // Lock the potential write of the dictionary to prevent multiple threads from creating the same application.
+            await _managedIdSemaphore.WaitAsync();
```

Comment:
I'd love to see tests at some level for checking correctness around concurrency.

Reply:
Solid point, is there a common way to do this I could use as a starting point?

Reply:
My initial thought would be to make several threads and set a gate so they all call the method at roughly the same time.

Reply:
sounds reasonable